execution-engine/src/wasi/common.rs

```diff
@@ -373,6 +395,189 @@ pub struct WasiWrapper {
     exit_code: Option<u32>,
     /// Whether clock functions (`clock_getres()`, `clock_gettime()`) should be enabled.
     pub(crate) enable_clock: bool,
+    /// Whether strace is enabled.
+    pub(crate) enable_strace: bool,
+}
+
+// Convert a vector of bytes into a printable string.
+// TODO: Perhaps we should use a more generous definition of "printable".
+fn strace_string(bytes: &Vec<u8>, max: usize) -> String {
+    let mut res = String::from("\"");
+    let n = if bytes.len() > max { max } else { bytes.len() };
+    for i in 0..n {
+        if 0x20 <= bytes[i] && bytes[i] < 0x7f {
+            if bytes[i] == b'\\' || bytes[i] == b'"' {
+                res.push_str("\\");
+            }
+            res.push_str(&String::from_utf8_lossy(&bytes[i..i + 1].to_vec()))
+        } else if bytes[i] == 9 {
+            res.push_str("\\t")
+        } else if bytes[i] == 10 {
+            res.push_str("\\n")
+        } else if i + 1 < n && b'0' <= bytes[i + 1] && bytes[i + 1] <= b'9' {
+            // The following character is a digit, so use three octal digits.
+            res.push_str(&format!("\\{:03o}", bytes[i]))
+        } else {
+            res.push_str(&format!("\\{:o}", bytes[i]))
+        }
+    }
+    res.push_str("\"");
+    if bytes.len() > max {
+        res.push_str("...")
+    }
+    res
+}
+
+fn strace_iovec<T: MemoryHandler>(
+    success: bool,
+    memory_ref: &mut T,
+    address: u32,
+    base: u32,
+    count: u32,
+) {
+    if !success {
+        eprint!("_");
+        return;
+    }
+    if let Ok(len) = memory_ref.read_u32(address) {
+        let max = if len < 32 { len } else { 32 } as usize;
+        // This inefficiently copies everything, but it's only used for tracing.
+        if let Ok(bufs) = (|| {
+            let iovecs = memory_ref.unpack_array::<IoVec>(base, count)?;
+            memory_ref.read_iovec_scattered_in_reverse(&iovecs)
+        })() {
+            let mut buf: Vec<u8> = Vec::new();
```

Comment:
Is this doing the same job as `.iter().flatten()`?

Reply:
This code has changed now for other reasons.